lib/spack/spack/compilers/__init__.py

```diff
@@ -970,7 +970,10 @@ def _default_make_compilers(cmp_id, paths):
     compilers = []
     for compiler_id, _, compiler in flat_compilers:
         make_compilers = getattr(compiler_id.os, "make_compilers", _default_make_compilers)
-        compilers.extend(make_compilers(compiler_id, compiler))
+        candidates = make_compilers(compiler_id, compiler)
+        # Discard candidates without a C compiler
+        candidates = [x for x in candidates if x.cc is not None]
```

Comment:
avoid the temporary, `extend` can take an iterator

Reply:
See https://github.com/spack/spack/pull/43778/commits/b8cc58ab22b86ee3f46af7aa1e54bd5fbff9a974